best_practices.md

```diff
@@ -0,0 +1,262 @@
+
+### Explanation
+Dynamic allocations can be costly in terms of performance, especially in functions that are called frequently. Using fixed-size containers like `std::array` or `std::tuple` can significantly reduce overhead. `ttnn::small_vector` provides a performance-optimized alternative to `std::vector` for most scenarios.
+
+### Motivation
+- **Performance**: Reduces the overhead associated with dynamic memory allocation.
+- **Predictability**: Improves predictability of memory usage and performance.
+- **Efficiency**: Enhances overall efficiency by leveraging fixed-size containers or optimized small vectors.
+
+### Example
+```
+// Allocating on heap
+std::vector<int> dynamicVector = {1, 2, 3, 4, 5};
+
+// No heap allocation
+std::array<int, 5> fixedArray = {1, 2, 3, 4, 5};
+```
+
+## 5. Avoid Using `.at()` for Vector Access
+
+### Practice
+Avoid calling the `.at()` function on vectors and other stl containers. Perform all necessary checks before using the vector and use `TT_ASSERT` or `TT_FATAL`. The `.at()` function throws exceptions without context, which can be frustrating for users. Overall, avoid STL calls that throw exceptions. All inputs should be validated on our side.
+
+### Explanation
+The `.at()` function performs bounds checking and throws an exception if the index is out of range. However, the exceptions lack detailed context, making them hard to debug. By using `TT_ASSERT`, `TT_FATAL` or `TT_THROW`, you can provide clearer error messages and avoid exceptions.
+
+### Motivation
+- **User Experience**: Provides more informative error messages, improving user experience.
+- **Debugging**: Easier to debug issues with clear assertions and fatal errors.
+- **Performance**: Avoids the overhead associated with exception handling.
+
+### Example
+```
+for (int i = 0; i < fp32_vec.size(); i++) {
+  fp32_vec.at(i) = rand_float() + offset; // Very bad, breaks a lot of optimiations. Additional check.
+}
+for (int i = 0; i < fp32_vec.size(); i++) {
+  fp32_vec[i] = rand_float() + offset; // Good!
+```
+
+
+## 6. Avoid `std::move` When Returning from Functions
+
+### Practice
+Don't use `std::move` when returning from a function, as it breaks Return Value Optimization (RVO).
+
+### Explanation
+RVO is an optimization technique that eliminates unnecessary copying or moving of objects when they are returned from a function. Using `std::move` can prevent the compiler from applying RVO, resulting in less efficient code.
+
+### Motivation
+- **Performance**: Ensures that RVO can be applied, minimizing unnecessary copying or moving of objects.
+- **Efficiency**: Maintains optimal performance by allowing the compiler to perform return value optimizations.
+
+### Example
+https://godbolt.org/z/enTqadjMz
+
+## 7. Avoid `const T&&` or `const auto&&` and Returning `const` Values
```

Comment:
Wdyt about splitting these?

Reply:
It is about overusing const.  But you are riggt const T&& and return const are different problems.